Wikipedia/Code/WMFArticleImageProtocol.m

```diff
@@ -55,37 +71,37 @@ - (void)startLoading {
     DDLogVerbose(@"Fetching image %@", self.request.URL);
     @weakify(self);
     [[WMFImageController sharedInstance] fetchImageWithURL:self.request.URL]
-    .thenInBackground(^id(WMFImageDownload* download) {
+    .thenInBackground(^WMFArticleImageProtocolResponseAndData*(WMFImageDownload* download) {
         @strongify(self);
         if(!self){
             return nil;
         }
-        
-        UIImage* image     = download.image;
-        NSString* mimeType = [self.request.URL wmf_mimeTypeForExtension];
-        NSData* data       = [image wmf_dataRepresentationForMimeType:mimeType serializedMimeType:&mimeType];
-        
-        NSURLResponse* response =
-        [[NSURLResponse alloc] initWithURL:self.request.URL
-                                  MIMEType:mimeType
-                     expectedContentLength:data.length
-                          textEncodingName:nil];
-        
-        return (response && data) ? @[response, data]: nil;
+        return [self getResponseAndDataForDownload:download];
     })
-    .then(^(NSArray* responseAndDataArray) {
+    .then(^id(WMFArticleImageProtocolResponseAndData* responseAndData) {
         @strongify(self);
         if(!self){
-            return;
+            return nil;
         }
 
-        if (!responseAndDataArray || (responseAndDataArray.count != 2)) {
-            return;
+        if (!responseAndData.response) {
```

Comment:
If there was any problem making a response or processing the image now returns errors so the catch block will fire. Formerly if anything went wrong in either of these parts "URLProtocolDidFinishLoading" was still invoked, which indicated success, when in fact the loading had failed, which is what the catch block correctly reports.